features/crc/soft/FastCRC.h

```diff
@@ -0,0 +1,202 @@
+                break;
+
+            case CRC_32BIT:
+                _polynomial = 0x04C11DB7;     // x32+x26+x23+x22+x16+x12+x11+x10+x8+x7+x5+x4+x2+x+1
+                _width = 32;
+                _inital_value = ~(0x0);
+                _final_xor  = ~(0x0);
+                _reflect_data = true;
+                _reflect_remainder = true;
+                break;
+        }
+    }
+
+    virtual ~FastCRC() { }
+
+    /** Initialize a CRC module, generate CRC static table
+     *
+     *  @return  0 on success or a negative error code on failure
+     */
+    virtual int32_t init() {
+        T remainder;
+        uint32_t polynomial = (_width < 8) ? (_polynomial << (8 - _width)) : _polynomial;
+
+        for (int32_t dividend = 0; dividend < 256; ++dividend) {
+            if ( _width < 8) {
+                remainder = dividend;
+            } else {
+                remainder = dividend << (_width - 8);
+            }
+            for (uint8_t bit = 8; bit > 0; --bit) {
+                if (remainder & TOP_BIT(_width)) {
+                    remainder = (remainder << 1) ^ polynomial;
+                } else {
+                    remainder = (remainder << 1);
+                }
+            }
+
+            /*
+             * Store the result into the table.
+             */
+            _crc_table[dividend] = remainder;
+        }
+        return 0;
+    }
+
+    /** Deinitialize a CRC module
+     *
+     * @return  0 on success, negative error code on failure
+     */
+    virtual int32_t deinit() {
+        return 0;
+    }
+
+    /** Get CRC polynomial
+     *
+     * Get the current CRC polynomial
+     *
+     * @return  Polynomial value
+     */
+    virtual uint32_t get_polynomial(void) const {
+        return _polynomial;
+    }
+
+    /** Compute CRC for the data input as per the polynomial set
+     *
+     *  @param  buffer  Data bytes
+     *  @param  size  Size of data
+     *  @param  crc  CRC
+     *  @return  0  on success or a negative error code on failure
+     *  @note CRC calculations are done byte wise.
+     */
+    virtual int32_t compute(void *buffer, data_size_t size, uint32_t *crc) {
+        MBED_ASSERT(crc != NULL);
+        MBED_ASSERT(buffer != NULL);
+
+        uint8_t *data = static_cast<uint8_t *>(buffer);
+        T remainder = _inital_value;
+        uint8_t data_byte;
+
+        for (data_size_t byte = 0; byte < size; byte++) {
+            if ( _width < 8 ) {
+                remainder = _crc_table[(data[byte] ^ remainder)];
+            } else {
+                data_byte = reflect_bytes(data[byte]) ^ (remainder >> (_width - 8));
+                remainder = _crc_table[data_byte] ^ (remainder << 8);
+            }
+        }
+        if ( _width < 8 ) {
+            remainder >>= (8 - _width);
+        }
+        *crc = (reflect_remainder(remainder) ^ _final_xor) & DATA_MASK(_width);
+        return 0;
+    }
+
+private:
+    uint32_t _polynomial;
+    uint8_t _width;
+    T _crc_table[256];
```

Comment:
Shouldn't the CRC table be same for same polynomial? In that case, have you thought of sharing the table for same polynomial types across multiple instances. That can help save lot of memory and will avoid computing the table for same polynomial type in init. May be the assumption here is to use same CRC object across the module(For E.g.: fs, block device), in that case should we protect the interfaces for thread-safety?

Reply:
Having common CRC table across modules will require proper thread-safety, but I am not sure if any application will have this use case. Can however think of it in future